oddbird/oddcontrast
# I want to share a url to a color combination

![](https://source.unsplash.com/featured/?cute,animal&share)

## Description
Allows users to share a URL with colors. Addresses #92 


## Steps to test/reproduce

1. [Open site](https://deploy-preview-97--oddcontrast.netlify.app/), and see that as you select colors, the URL hash updates with the current values.
2. Reload the page and see that the colors persist
3. Copy the URL into a new tab, browser, etc, and see that the colors persist
4. Verify that changing the slider positions doesn't create new History steps (it doesn't make sense to go back or forward between substeps)
5. Check different formats, especially `hex`, where the display format is different from the color space- [Example](https://deploy-preview-97--oddcontrast.netlify.app/#hex__rgb(7.2895~_95~_46.707~)__rgb(19.1~_28.8~_75.837~))
7. I opted to make the URL readable over "correct", so I replaced spaces with `_` rather than `%20` and replaced `%` with `~`.


Some things to note/questions-

1. There is a blink of our default colors while the colors are loaded from the URL. This is because we are rendering the page on the server, where we don't have the hash, and then applying the changes after the page is loaded. We could disable SSR, but that would impact performance for everyone. I'm open to other ideas to fix this.
2. Check out of gamut color inputs by changing the URL- [Example with color(display-p3 2 1 1)](https://deploy-preview-97--oddcontrast.netlify.app/#p3__color(display-p3_2_1_1)__color(display-p3_0.2119_0.2854_0.7308)) - these will load as out of gamut colors, while the URL immediately gets re-mapped into gamut. Should we instead map them to in gamut, in the same way we handle conversions? (Answer: yes, this is now mapped)
3. If any of the values are missing or invalid from the URL, no values will be changed, the default color will be loaded, and the URL will be replaced with the values of the default colors. We don't show an alert that it was invalid. There's some potential for loss of user input, but since they're likely copying and pasting it from somewhere, or clicking a link, the URL bar shouldn't be their only record of the color.
4. It currently is updating the URL on every input. We could debounce it, but I'm not sure it's necessary. (This is now debounced)
